mia/commands/definition.py

```diff
@@ -212,103 +207,144 @@ def create_apps_lock_file():
+            # Use the default repository if no repo has been provided.
+            if 'repo' not in app_info:
+                app_info['repo'] = settings['defaults']['repository_id']
+
+            # Use the latest application version code.
+            if handler.args['--force-latest'] or 'code' not in app_info:
+                app_info['code'] = 'latest'
+
+            # Get the application info.
+            app_info = _xml_get_app_lock_info(repositories_data, app_info)
+            if app_info is not None:
+                repo_name = repositories_data[app_info['repository_id']]['name']
+                msg = ' - found `%s` in the %s repository.'
+                print(msg % (app_info['package_name'], repo_name))
+                apps_list.append(app_info)
+                continue
+
+        warnings_found = True
+
+    # Give the user a chance to fix any possible errors.
+    if warnings_found and not input_confirm('Warnings found! Continue?'):
+        sys.exit(1)
 
-        app_info['package_name'] = "%s" % app_package_name
-        app_info['package_url'] = "%s/%s" % \
-                                  (repo_info['base_url'], app_package_name)
+    return apps_list
 
-        if app_info['code'] == 'latest':
-            app_info['code'] = int(app_version_code)
 
-        print(' - found: %s:%s' % (app_info['name'], app_info['code']))
+def _xml_get_app_lock_info(data, app_info):
+    app_name = None
+    app_package_name = None
+    app_version_code = None
 
-    if warnings_count and not input_confirm('Warnings found! Continue?'):
-        sys.exit(1)
+    # Prepare a list of repositories to look into.
+    repositories = [app_info['repo']]
+    if 'fallback' in data[app_info['repo']]:
+        repositories.append(data[app_info['repo']]['fallback'])
 
-    return repo_apps
+    for repo in repositories:
+        for tag in data[repo]['tree'].findall('application'):
+            if tag.get('id') and tag.get('id') == app_info['name']:
+                app_name, app_package_name, app_version_code = \
+                    _xml_get_app_download_info(tag, app_info['code'])
 
+        # Only try the fallback repository if the application was not found.
+        if app_package_name is not None:
+            break
 
-def _xml_get_application_tag(tree_root, target):
-    for tag in tree_root.findall('application'):
-        if tag.get('id') and tag.get('id') == target:
-            return tag
+    if app_package_name is None and app_info['code'] == 'latest':
+        print(' - no such app: %s' % app_info['name'])
+        return None
+    elif app_package_name is None:
+        print(' - no package: %s:%s' % (app_info['name'], app_info['code']))
+        return None
 
-    return None
+    return {
+        'name': app_name,
+        'repository_id': repo,
+        'package_name': app_package_name,
+        'package_code': int(app_version_code),
+        'package_url': urljoin(data[repo]['url'], app_package_name),
```

Comment:
urljoin is not only python3, but depending on whether repo url has trailing slash, it can easily strip the last bit of the url path:

```
In [1]: from urllib.parse import urljoin

In [2]: url = 'https://f-droid.org/repo'

In [3]: file = 'org.fdroid.fdroid_830.apk'

In [4]: urljoin(url, file)
Out[4]: 'https://f-droid.org/org.fdroid.fdroid_830.apk'

In [5]: url = 'https://f-droid.org/repo/'

In [6]: urljoin(url, file)
Out[6]: 'https://f-droid.org/repo/org.fdroid.fdroid_830.apk'

In [7]: 
```